File: 프로그래머스/3/118668. 코딩 테스트 공부/코딩 테스트 공부.py
from heapq import heappush, heappop


def solution(alp, cop, problems):
    problems += [[0, 0, 1, 0, 1], [0, 0, 0, 1, 1]]

    max_alp = max(problem[0] for problem in problems)
    max_cop = max(problem[1] for problem in problems)

    q = [(0, (alp, cop))]
    checked = {(alp, cop): 0}
    limit = 150

    while q[0][1][0] < max_alp or q[0][1][1] < max_cop:
        time, (alp, cop) = heappop(q)

        for alp_req, cop_req, alp_rwd, cop_rwd, cost in problems:
            if alp < alp_req or cop < cop_req:
                continue

            next_cost = time + cost
            next_alp = min(alp + alp_rwd, limit)
            next_cop = min(cop + cop_rwd, limit)

            if (next_alp, next_cop) in checked and checked[(next_alp, next_cop)] <= next_cost:
                continue
            
            checked[(next_alp, next_cop)] = next_cost
            heappush(q, (next_cost, (next_alp, next_cop)))

    return heappop(q)[0]

File: 프로그래머스/3/118668. 코딩 테스트 공부/코딩 테스트 공부.java
import java.util.*;

class Solution {
    
    public int solution(int alp, int cop, int[][] problems) {
        int n = problems.length;
        int[][] newProblems = new int[n + 2][5];
        for (int i = 0; i < n; i++) {
            newProblems[i] = problems[i];
        }
        newProblems[n] = new int[]{0, 0, 1, 0, 1};
        newProblems[n + 1] = new int[]{0, 0, 0, 1, 1};

        int maxAlp = 0, maxCop = 0;
        for (int[] p : newProblems) {
            maxAlp = Math.max(maxAlp, p[0]);
            maxCop = Math.max(maxCop, p[1]);
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        Map<String, Integer> checked = new HashMap<>();

        pq.offer(new int[]{0, alp, cop});
        checked.put(alp + "," + cop, 0);

        int limit = 150;

        while (true) {
            int[] cur = pq.poll();
            int time = cur[0];
            int curAlp = cur[1];
            int curCop = cur[2];

            if (curAlp >= maxAlp && curCop >= maxCop) {
                return time;
            }

            for (int[] p : newProblems) {
                int alpReq = p[0], copReq = p[1];
                int alpRwd = p[2], copRwd = p[3], cost = p[4];

                if (curAlp < alpReq || curCop < copReq) continue;

                int nextCost = time + cost;
                int nextAlp = Math.min(curAlp + alpRwd, limit);
                int nextCop = Math.min(curCop + copRwd, limit);

                String key = nextAlp + "," + nextCop;

                if (checked.containsKey(key) && checked.get(key) <= nextCost) continue;

                checked.put(key, nextCost);
                pq.offer(new int[]{nextCost, nextAlp, nextCop});
            }
        }
    }
}